athena/models/fastspeech.py

```diff
@@ -292,36 +294,55 @@ def inference(self, phoneme_sequences, duration_sequences, alpha=1.0):
         # the duration value of one phoneme at least is one
-        total_durations = tf.reduce_sum(duration_sequences, axis=1) # [batch]
-        max_duration = tf.reduce_max(total_durations)
-
-        def expand_phoneme(batch_i):
-            phoneme_sequence = phoneme_sequences[batch_i]
-            duration_sequence = duration_sequences[batch_i]
-            total_duration = total_durations[batch_i]
-            expanded_array = tf.TensorArray(tf.float32, size=0, dynamic_size=True)
-            expanded_array = expanded_array.write(0, tf.zeros([d_model]))
-            for step_i in tf.range(x_steps):
-                duration = duration_sequence[step_i]
-                if duration == 0:
-                    continue
-                phoneme = phoneme_sequence[step_i]
-                expanded_phoneme = tf.expand_dims(phoneme, axis=0)
-                expanded_phoneme = tf.tile(expanded_phoneme, [duration, 1]) # [duration, d_model]
-                expanded_array = expanded_array.unstack(tf.concat([expanded_array.stack(),
-                                                                   expanded_phoneme], axis=0))
-            expanded_array = tf.concat([expanded_array.stack(),
-                                        tf.zeros([max_duration - total_duration, d_model])],
-                                       axis=0)[1:] # [max_duration, d_model]
-            return expanded_array
+
+        # max_duration_i: the max duration of all the indexes
+        max_duration_i = tf.reduce_max(duration_sequences)
+        background = tf.ones([batch, x_steps, max_duration_i], dtype=tf.int32) * -1
+
+        # durations represents frame size for each index, shape: [batch, x_steps, max_duration_i]
+        durations = tf.tile(duration_sequences[:, :, tf.newaxis], [1, 1, max_duration_i])
+        duration_order_array = tf.range(max_duration_i)
+        duration_order_array = tf.tile(duration_order_array[tf.newaxis, tf.newaxis, :],
+                                       [batch, x_steps, 1])
+        step_order_array = tf.range(x_steps)
+        step_order_array = tf.tile(step_order_array[tf.newaxis, :, tf.newaxis],
+                                   [batch, 1, max_duration_i])
+        #duration_indexes examples:
+        #if we discard the batch dimension, duration_indexes will be like:
+        #    [[0, 0, -1, -1],
+        #     [1, 1, 1, -1],
+        #     [2, -1, -1, -1]]
+        #And the corresponding duration sequence is [2, 3, 1]
+        duration_indexes = tf.where(duration_order_array < durations, step_order_array, background)
+        duration_indexes = tf.reshape(duration_indexes, [batch, -1]) # [batch, x_steps*max_duration_i]
+
+        valid_durations = tf.cast((duration_indexes != -1), dtype=tf.int32)
+        total_duration_batches = tf.reduce_sum(valid_durations, axis=-1) # [batch]
+        y_steps = tf.reduce_max(total_duration_batches)
+
+        def validate_duration_sequences(batch_i):
```

Comment:
we should add some comments for this function, why do we need to validate duration seq? Is it because num of input frames is different from total duration?

Reply:
the name of the function is a little weird considered its function. more comments would be added.